Fix perfmon for qt changes, add counters, cleanup

pwinston opened the pull request:
# Description
* Fix perfmon from Qt re-org,.
* Added new perfmon `add_counter_event()` API.
* Lots of docstring cleanup and type fixes.
* Fix `PerfTraceFile`to only write at the end:
    * We used to write _as we went_, and some stuff was legacy from that era.
* Add one unit test: `test_perf.py`
   * Add pytest flag `--perfmon-only`
   * Add pytest mark `@pytest.mark.perfmon`
   * Add new Cirrus.yml test pass for this one perfmon test

We need a seperate test pass so we can define `NAPARI_PERFTEST` which is only checked once on module import.

## Type of change
- [x] Bug file (Qt issue)
- [x] New Features: `add_counter_event()`
- [x] New unit test

# How has this been tested?
- [x] Manual testing
- [x] The one new unit test

## Final checklist:
- [x] I have commented my code, particularly in hard-to-understand areas

# Example of add_counter_event()

A counter displays as a bar graph or stacked bar graph (if multiple values). For instance here it's showing the length of the delay queue. Basically it's visibility into what scalar values are doing over time.

<img width="1022" alt="perfmon-counters" src="https://user-images.githubusercontent.com/4163446/90286168-4cbf4500-de43-11ea-8812-34ca8c13550e.png">

pwinston committed:
Perfmon changes from 1510

pwinston committed:
Better comments

pwinston committed:
Better file docstring and remove unnecessary stuff

sofroniewn commented on napari/utils/perf/_event.py:
```diff
@@ -2,7 +2,14 @@
+from collections import namedtuple
+from typing import Dict, Optional
+
+# A span of time for an event.
```
I really like this use of namedtuple, blog your post here https://tobeva.com/articles/brain-compatible-code/ really resonated with me!!

sofroniewn reviewed (approved):
This looks good to me, thanks for the small digestible PR. Really nice to have this good tooling in for the perfmon!

pwinston committed:
Comments and cleanup

pwinston committed:
Fix for qt file re-org

pwinston committed:
Better file docstring

pwinston committed:
None Types

pwinston committed:
Add type

pwinston committed:
Slim/improve doc string

In light changes to napari.utils.perf.__init__.py docs.

jni reviewed (commented):
@pwinston thanks for the fix! My main Q is: could we add a test that checks that this doesn't break? Otherwise it's bound to happen again and again!

pwinston committed:
Merge branch 'master' into feature/perfmon-counters

pwinston committed:
Initial test, does not work yet

pwinston committed:
Add new --perfmon-only flag to pytest

pwinston committed:
Rename config, add path

pwinston committed:
use explicit start/stop

No good way to do it on startup

pwinston committed:
Fix format?

pwinston committed:
Add comment

pwinston committed:
Make file write only on exit, fix test

pwinston committed:
Fix test skipping logic

pwinston committed:
Not working for some readon

pwinston committed:
Remove flags from test coverage?

pwinston committed:
Does this need quotes?

pwinston committed:
Add FLAGS variable

pwinston committed:
No code coverage for perfmon

pwinston committed:
Different style conditional

pwinston committed:
Fix env var name!

pwinston committed:
Don't delete self._qt_window twice

pwinston committed:
Make helper private

jni commented on napari/conftest.py:
```diff
@@ -37,15 +37,22 @@ def fetch(data_name):
-    the pytest command) to use this flag.
+    --show-viewer
+        Show viewers during tests, they are hidden by default. Showing viewers
+        decreases test speed by about %18.
```
```suggestion
        decreases test speed by about 18%.
```
? =) Also, maybe we can just round it to 20%? "decreases test speed by approximately 18.42927%" 😂

pwinston replied:
That was not my comment, Talley's I think, but I changed it to say "by about 20%".

jni commented on napari/conftest.py:
```diff
@@ -37,15 +37,22 @@ def fetch(data_name):
+    Notes
+    -----
+    Due to the placement of this conftest.py file, you must specifically name
+    the napair folder such as "pytest napari --show-viewer"
```
```suggestion
    the napari folder such as "pytest napari --show-viewer"
```

pwinston replied:
Fixed

jni commented on napari/utils/perf/_event.py:
```diff
@@ -16,34 +23,41 @@ class PerfEvent:
-    category :str
-        Comma separated categories such has "render,update".
+    phase : str
+        The Chrome Tracing "phase" such as "X", "I", "C".
```
It seems you've deleted the little table of meanings for these — worth bringing it back! Specifically this description was very useful:

     phase : str
         The chrome://tracing "phase" (event type). The spec defines
         around 20 phases we only support two right now:
              "X" - Complete Events
              "I" - Instant Events

pwinston replied:
Added back the table.

jni commented on napari/utils/perf/_event.py:
```diff
@@ -54,39 +68,76 @@ def __init__(
     def __init__(self, name: str, time_ns: int, **kwargs):
-        super().__init__(name, time_ns, time_ns, **kwargs)
-        self.phase = "I"  # instant event
+        # Instant events are phase "I" in their trace space.
```
Is this supposed to say "their trace space" or "the trace spec"?

pwinston replied:
"Spec" but I just removed that line. I think it's redundant with the above table added back.

jni commented on napari/utils/perf/_timers.py:
```diff
@@ -83,16 +74,34 @@ def add_event(self, event: PerfEvent):
+        -----
+        For example add_counter_event("draw", triangles=5, squares=10).
+        """
+        self.add_event(CounterEvent(name, perf_counter_ns(), **kwargs))
```
It seems to me that these classes are a rather heavyweight way to express this code — would this not be equally simple?

```suggestion
        self.add_event(PerfEvent(name, perf_counter_ns(), phase='C', **kwargs))
```

and so on for each supported phase?

pwinston replied:
I was going to say I like `CounterEvent` on that line and not just `C`. But yes the sub-classes were not at all pulling their weight, so I got rid of them. One thing they were doing is Instant and Counter events had only one time, the time it happened, and not both a start and an end time.. But not really worth it just for that. I just pass "now" in as the start and end now.

jni commented on napari/utils/perf/_timers.py:
```diff
@@ -159,17 +197,18 @@ def perf_timer(
-    # zero it's best practice not to commit perf_timers, think of
-    # them like debug prints.
+    # create a do-nothing context object. Not zero overhead but
+    # very small.
```
I think the previous message was more informative and useful, is there a reason you removed it?

pwinston replied:
I think "think of them as debug prints" is a higher level piece of advice, that's not part of describing what the context object actually does. This higher level advice is now part of the overall perfmon doc string in `utils.perf.__init__`.

jni reviewed (approved):
Thanks @pwinston! I've left some minor comments — mostly around comments! Please fix the documentation bits before merging, where I think there were a couple of regressions. I'd also love the extra class definitions removed but that can happen later.

kevinyamauchi commented on napari/utils/perf/__init__.py:
```diff
@@ -1,41 +1,42 @@
 
-USE_PERFMON is true only if NAPARI_PERFMON environment variable is set and not
-zero.
+To enable perfmon define the env var NAPARI_PERFMON to be non-zero.
```
I think it would be nice to add a bit more detail on what is meant by "non-zero". Perhaps something like:

```suggestion
To enable perfmon define the env var NAPARI_PERFMON to be non-zero. For example:

- set to 1 to activate perfmon and use the menu Debug -> Performance Trace in the napari GUI
- set to the path to your perfmon config file to configure napari to start up with perfmon set up as specified in the config file. See below for more information of the perfmon config file.
```

pwinston replied:
I thought that was covered somewhere, buy couldn't find where. I redid that whole comment block just now and included your ideas. Seems better now.

kevinyamauchi replied:
Awesome! Thanks!

kevinyamauchi reviewed (approved):
This looks good to me! Stoked to have the qt tracing back. I left just have a minor comment below. I assume we will eventually have a docs page for this, but it would be nice to have a bit more detail on how to activate/configure perfmon in the meantime.

pwinston committed:
Review comments

pwinston committed:
Adde back "around 20%"

pwinston committed:
Remove PerfEvent subclasses

They were not doing enough to justify
their existence. Thought they would
do more, but didn't end up happening.
